### ISSUE TITLE
test_fakepwd.PwdModuleTests.test_noSuchUID fails on RHEL6

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6201]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)| @tomprince reported|
|-|-|
|Trac ID|trac#6201|
|Type|defect|
|Created|2012-12-04 16:30:07Z|
|Branch|https://github.com/twisted/twisted/tree/uid-overflow-6201|

https://buildbot.twistedmatrix.com/builders/rhel6-x86_64-py2.6/builds/13
```
[FAIL]
Traceback (most recent call last):
  File "/srv/buildslave/slave-fedora-rhel6-x86_64/rhel6-x86_64-py2_6/Twisted/twisted/internet/defer.py", line 138, in maybeDeferred
    result = f(*args, **kw)
  File "/srv/buildslave/slave-fedora-rhel6-x86_64/rhel6-x86_64-py2_6/Twisted/twisted/internet/_utilspy3.py", line 41, in runWithWarningsSuppressed
    reraise(exc_info[1], exc_info[2])
  File "/srv/buildslave/slave-fedora-rhel6-x86_64/rhel6-x86_64-py2_6/Twisted/twisted/internet/_utilspy3.py", line 37, in runWithWarningsSuppressed
    result = f(*a, **kw)
  File "/srv/buildslave/slave-fedora-rhel6-x86_64/rhel6-x86_64-py2_6/Twisted/twisted/python/test/test_fakepwd.py", line 59, in test_noSuchUID
    self.assertRaises(KeyError, self.database.getpwuid, -13)
  File "/srv/buildslave/slave-fedora-rhel6-x86_64/rhel6-x86_64-py2_6/Twisted/twisted/trial/_synctest.py", line 335, in assertRaises
    failure.Failure().getTraceback()))
twisted.trial.unittest.FailTest: <type 'exceptions.OverflowError'> raised instead of KeyError:
 Traceback (most recent call last):
  File "/srv/buildslave/slave-fedora-rhel6-x86_64/rhel6-x86_64-py2_6/Twisted/twisted/trial/_asynctest.py", line 106, in _run
    utils.runWithWarningsSuppressed, self._getSuppress(), method)
  File "/srv/buildslave/slave-fedora-rhel6-x86_64/rhel6-x86_64-py2_6/Twisted/twisted/internet/defer.py", line 138, in maybeDeferred
    result = f(*args, **kw)
  File "/srv/buildslave/slave-fedora-rhel6-x86_64/rhel6-x86_64-py2_6/Twisted/twisted/internet/_utilspy3.py", line 37, in runWithWarningsSuppressed
    result = f(*a, **kw)
  File "/srv/buildslave/slave-fedora-rhel6-x86_64/rhel6-x86_64-py2_6/Twisted/twisted/python/test/test_fakepwd.py", line 59, in test_noSuchUID
    self.assertRaises(KeyError, self.database.getpwuid, -13)
--- <exception caught here> ---
  File "/srv/buildslave/slave-fedora-rhel6-x86_64/rhel6-x86_64-py2_6/Twisted/twisted/trial/_synctest.py", line 328, in assertRaises
    result = f(*args, **kwargs)
exceptions.OverflowError: user id is not in range(-1, 2^32-1)


twisted.python.test.test_fakepwd.PwdModuleTests.test_noSuchUID
```

<details><summary>Searchable metadata</summary>

```
trac-id__6201 6201
type__defect defect
reporter__tom_prince tom.prince
priority__normal normal
milestone__ 
branch__branches_uid_overflow_6201 branches/uid-overflow-6201
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1354638607000000 1354638607000000
changetime__1358062085000000 1358062085000000
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

[#6208](https://github.com/twisted/twisted/issues/6208) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

(In [36770]) Branching to 'uid-overflow-6201'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/uid-overflow-6201).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

[#6208](https://github.com/twisted/twisted/issues/6208) was not a duplicate; [#6248](https://github.com/twisted/twisted/issues/6248) was.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @glyph|
|-|-|

[#6248](https://github.com/twisted/twisted/issues/6248) has a link to http://bugs.python.org/issue4591 which has a patch attached changing the overflow behavior of `getpwuid` to raise `OverflowError`.  However, the patch has been applied to no upstream version of CPython (including any of their development branches).

Only RHEL seems to be shipping a Python with this behavior.  The RHEL6 Python source can be found at <ftp://ftp.redhat.com/pub/redhat/linux/enterprise/6Server/en/os/SRPMS/python-2.6.6-29.el6.src.rpm>, including the patch file python-2.6.6-uid-gid-overflows.patch which changes a bunch of stuff around.

This seems like a bad choice on the part of RedHat - changing the behavior of a public API when upstream hasn't even accepted the proposed change, and at this point will never accept it for Python 2.6 (the version shipped with RHEL).  The result is exactly what we see here, an incompatibility with every other platform/Python combination, breaking things.

The change in the branch fixes the test but doesn't fix the problem: our verified fake never produces `OverflowError`, but any code that wants to work on RHEL6 will have to handle `OverflowError` instead of `KeyError`.  This lessens the value of the fake, and seems like an end run around what the tests are actually trying to verify rather than a true fix.

Alternate solutions, in no particular order:

  1. RedHat should stop shipping incompatible changes to public interfaces that necessitate specific support code for their platform in otherwise platform-agnostic code.
  1. Twisted should stop supporting RHEL6 so that it doesn't matter if Twisted works there or not.
  1. Somehow detect this case (call the function with a negative number and observe `OverflowError`?) and skip the test.  This leaves the fake broken on RHEL6 but makes the test suite green, giving the appearance that all is well (which, frankly, may be all anyone cares about).
    1. The above, but file a ticket for fixing it for real later.
  1. Somehow detect this case and change the test to require `OverflowError` _instead_ of `KeyError` (rather than accepting either) and then fix the implementation to pass the new version of the test.
    1. The above, plus introduce a new test which always exercises the `KeyError` case and somehow finds a non-overflowing UID to use to do so (iterate over a big UID range and hope that something in it is not allocated?).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph removed owner|
|-|-|

Replying to [exarkun](#note_5):

>     1. The above, plus introduce a new test which always exercises the `KeyError` case and somehow finds a non-overflowing UID to use to do so (iterate over a big UID range and hope that something in it is not allocated?).

I opted to do this instead, since the intent here is to test the "user does not exist" case, and that case exists on Red Hat's funky patched version as well.  It's also possible to get a TypeError from this API, but we're not testing that.  Unfortunately it's not formally required that a UID from 999 to 1 has to be unallocated, but hey, it's true on all of our supported platforms.  (And those UIDs are reserved for the platform by convention, so, there you go.)

[Forced a new build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/uid-overflow-6201), and just to make it easier, [here's the patch too](http://twistedmatrix.com/~diffresource.twistd/6201).

Perhaps I should also file a ticket for testing the other edge cases, and adding support for them to the fake?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @glyph|
|-|-|

Thanks.  This looks like a saner change.  I like the idea of filing a ticket for covering some other edge cases, too.  Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

[#6254](https://github.com/twisted/twisted/issues/6254) covers the other change.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set status to `closed`|
|-|-|

(In [36805]) Merge uid-overflow-6201

Author: glyph

Reviewer: exarkun

Fixes: [#6201](https://github.com/twisted/twisted/issues/6201)

Address test failures on Red Hat buildbot; change the verified-fake tests for
the pwd module to detect a value for a non-existent UID that is still in the
uid_t range, for Red Hat's bundled patch.
